### PR TITLE
Adaptation for the upgrade via SMT (bsc#1094865)

### PR DESCRIPTION
**Pull request by lslezak**
- This PR contains adaptations needed for upgrade via SMT
- See bug https://bugzilla.suse.com/show_bug.cgi?id=1094865
- Still WIP...

### CODE REVIEW

File: src/lib/registration/sw_mgmt.rb [hunk 1]
```diff
@@ -443,18 +444,22 @@ def self.copy_old_credentials(source_dir)
...
       end
 
-      # check for NCC credentials
-      ncc_file = File.join(source_dir, dir, "NCCcredentials")
-      copy_old_credentials_file(ncc_file)
+      Dir[File.join(source_dir, dir, "*")].each do |path|
+        # skip non-files or NCCcredentials, they are handled differently
+        next if !File.file?(path) || File.basename(path) == "NCCcredentials"
+
+        new_path = File.join(dir, File.basename(path))
```
Comment (jreidinger) [inline src/lib/registration/sw_mgmt.rb]: what about merging those two cases? like
```
Dir[File.join(source_dir, dir, "*")].each do |path|
  # skip non-files
   next if !File.file?(path)

   new_path =  File.basename(path) == "NCCcredentials" ? SUSE::Connect::YaST::GLOBAL_CREDENTIALS_FILE : File.join(dir, File.basename(path))
```

File: test/sw_mgmt_spec.rb [hunk 2]
```diff
@@ -225,30 +226,44 @@
...
+      expect(File).to receive(:exist?).with(ncc_credentials).and_return(false)
+
+      expect(subject).to receive(:`).with("cp -a " + scc_credentials + " " +
+        File.join(target_dir, "SCCcredentials"))
       expect(SUSE::Connect::YaST).to receive(:credentials).and_return(OpenStruct.new)
 
       expect { subject.copy_old_credentials(root_dir) }.to_not raise_error
     end
+
+    it "copies old SCC credentials at upgrade" do
```
Comment (jreidinger) [inline test/sw_mgmt_spec.rb]: copy and paste error, two identical states. I think rspec will complain.

File: test/sw_mgmt_spec.rb [hunk 3]
```diff
@@ -225,30 +226,44 @@
...
+        .and_return([smt_credentials])
+      expect(File).to receive(:exist?).with(smt_credentials).and_return(true)
+      expect(File).to receive(:exist?).with(ncc_credentials).and_return(false)
+
+      expect(subject).to receive(:`).with("cp -a " + smt_credentials + " " +
+        File.join(target_dir, "SMT-http_smt_example_com"))
+      expect(SUSE::Connect::YaST).to receive(:credentials).and_return(OpenStruct.new)
+
+      subject.copy_old_credentials(root_dir)
+    end
```
Comment (jreidinger) [inline test/sw_mgmt_spec.rb]: and if you want to be so explicit, add also test for RMT.

Comment (lslezak) [reply test/sw_mgmt_spec.rb]: Um, that's not necessary, both RMT and SMT use dynamic name of the credentials (usually the host name). So it's enough to test anything different to the standard `SCCcredentials`.

**Comment by lslezak** [comment]
Unfortunately that `private_class_method def self.foo` trick does not work, yardoc cannot parse this and fails. So Travis fails as well... :worried:

File: src/lib/registration/ssl_certificate.rb [hunk 4]
```diff
@@ -175,12 +191,15 @@ def import_to_system
     # to installation system is slightly different to the one followed
     # to import certificates to a installed system.
     #
+    # @param target_path [String] where the imported certificate will be saved,
+    #   the pathe should contain the INSTSYS_CERT_DIR prefix otherwise it might
```
Comment (jreidinger) [inline src/lib/registration/ssl_certificate.rb]: typo pathe

File: src/lib/registration/ssl_certificate.rb [hunk 5]
```diff
@@ -175,12 +191,15 @@ def import_to_system
...
+    #   not work correctly.
     # @return [Boolean] true if import was successful; false otherwise.
     #
     # @see update_instsys_ca
-    def import_to_instsys
+    def import_to_instsys(target_path = self.class.default_certificate_path)
```
Comment (jreidinger) [inline src/lib/registration/ssl_certificate.rb]: to be honest I do not like much this new param as it is confusing when seeing it in use. consider `certificate.import_to_instsys(path)` this for me looks like it imports from path and not to path. So what about named param `to`? like `def import_to_instsys(to: self.class.default_certificate_path)` so usage will looks like `certificate.import_to_instsys(to: path)` ?

File: src/lib/registration/sw_mgmt.rb [hunk 6]
```diff
@@ -443,18 +444,22 @@ def self.copy_old_credentials(source_dir)
         ::FileUtils.mkdir_p(dir)
       end
 
-      # check for NCC credentials
-      ncc_file = File.join(source_dir, dir, "NCCcredentials")
-      copy_old_credentials_file(ncc_file)
+      Dir[File.join(source_dir, dir, "*")].each do |path|
```
Comment (jreidinger) [inline src/lib/registration/sw_mgmt.rb]: note that this by default skip hidden files starting with "." is it problem here?

File: src/lib/registration/sw_mgmt.rb [hunk 7]
```diff
@@ -443,18 +444,22 @@ def self.copy_old_credentials(source_dir)
...
+        end
 
-    def self.copy_old_credentials_file(file)
-      return unless File.exist?(file)
+        copy_old_credentials_file(path, new_path)
+      end
+    end
 
-      new_file = SUSE::Connect::YaST::GLOBAL_CREDENTIALS_FILE
+    private_class_method def self.copy_old_credentials_file(file, new_file)
```
Comment (jreidinger) [inline src/lib/registration/sw_mgmt.rb]: it is used also here. Here it does not cause problem?

Comment (lslezak) [reply src/lib/registration/sw_mgmt.rb]: No, because there is no yardoc comment... :confused:

File: src/lib/registration/sw_mgmt.rb (hunk 6 above)
Comment (lslezak) [reply src/lib/registration/sw_mgmt.rb]: Hidden credentials file would be very strange, I have never seen that.

File: src/lib/registration/ssl_certificate.rb (hunk 5 above)
Comment (lslezak) [reply src/lib/registration/ssl_certificate.rb]: Um, the usual usage looks like
```ruby
cert = SslCertificate.load_file(cert_file)
cert.import_to_instsys(target_path)
```

Where it is IMHO obvious what is loaded from and saved to.

I do not like adding a parameter name if there is only one parameter. If the name is confusing we could change the method name to `save_to_instsys(path)` or something like that.

Comment (jreidinger) [reply src/lib/registration/ssl_certificate.rb]: well, method `save_to_instsys`  makes good sense for me.

File: src/lib/registration/sw_mgmt.rb (hunk 6 above)
Comment (jreidinger) [reply src/lib/registration/sw_mgmt.rb]: ok, fine. I just want to point it out.

File: src/lib/registration/sw_mgmt.rb (hunk 7 above)
Comment (jreidinger) [reply src/lib/registration/sw_mgmt.rb]: ok

File: src/lib/registration/ssl_certificate.rb (hunk 5 above)
Comment (jreidinger) [reply src/lib/registration/ssl_certificate.rb]: but just NP, so feel free to merge.

Comment (jreidinger) [reply src/lib/registration/ssl_certificate.rb]: btw, here I do not get whole import name, when you actually exporting certificate and do not import anything.

**Pull request merged**